modules/core/nuxeo-core-binarymanager-cloud/nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/blob/s3/S3BlobStore.java

```diff
@@ -367,15 +373,23 @@ protected void clearBucket() {
 
     @Override
     public boolean readBlob(String key, Path dest) throws IOException {
+        ByteRange byteRange;
+        if (allowByteRange) {
+            MutableObject<String> keyHolder = new MutableObject<>(key);
+            byteRange = getByteRangeFromKey(keyHolder);
+            key = keyHolder.getValue();
+        } else {
+            byteRange = null;
+        }
         String objectKey;
         String versionId;
-        int seppos = key.indexOf(VER_SEP);
-        if (seppos < 0) {
-            objectKey = key;
-            versionId = null;
-        } else {
+        int seppos;
+        if (useVersion && (seppos = key.indexOf(VER_SEP)) > 0) {
```

Comment:
What do you think about introducing an object for key which does all the needed parsing (byte range and version) and holds methods such as `containsRange`.

Reply:
You mean a helper? In the vast majority of cases the key is just a string without structure. It's only for S3 versions and byte ranges that we need additional semantics but I don't want to overhaul the whole BlobStore API to add an abstraction here, which is why I did this string parsing.

Reply:
I meant a helper, more a data object with convenient methods and a static parse method (for instance). We would use only here and where we need to read range and not to use in a method signature as the change would be very big and not relevant in all cases (as you said).

`VER_SEP` could be a blocker for such helper as it is only for S3 and we don't want it in a "generic" thing.